### ISSUE TITLE
Race condition in PoolableLifestyleManager.CreatePool

### ISSUE THREAD

**Issue by krinmik** [castleproject/Windsor#72]
This was already detected in http://issues.castleproject.org/youtrack/issue/IOC-381.

Further analysis shows, that parallel Execution on CreatePool will fail with a ComponentRegistrationException. A poosible Bugfix would be to introduce a static lock Object:

```
private static object poolFactoryLock = new object();
protected IPool CreatePool(int initialSize, int maxSize)
{
    if (!Kernel.HasComponent(typeof(IPoolFactory)))
    {
        lock (poolFactoryLock)
        { 
            if (!Kernel.HasComponent(typeof(IPoolFactory))) // Avoid the race
            {
                Kernel.Register(Component.For<IPoolFactory>().ImplementedBy<DefaultPoolFactory>().NamedAutomatically("castle.internal-pool-factory"));
            }
        }
    }
}
```

I did some unittest to Repro it (PooledLifestyleManagerTestCase):

```
        [Test]
        public void CreatePoolRaceCondition()
        {
            using (AutoResetEvent evt = new AutoResetEvent(false))
            {
                MyKernel kernel = new MyKernel(evt);

                var manager1 = new MyClass();
                var manager2 = new MyClass();

                manager1.Init(null, kernel, null);
                manager2.Init(null, kernel, null);

                Task.Factory.StartNew(() => manager1.CreatePool());
                Task.Factory.StartNew(() => manager2.CreatePool());

                Thread.Sleep(TimeSpan.FromSeconds(5));
                evt.Set();
                Thread.Sleep(TimeSpan.FromSeconds(5));

                Assert.That(kernel.parallelCount, Is.EqualTo(0));
            }
        }

        private class MyClass : PoolableLifestyleManager
        {
            public MyClass()
                : base(1, 2)
            {            
            }

            public void CreatePool()
            {
                base.CreatePool(2, 5).Dispose();
            }
        }

        private class MyKernel : IKernel
        {
            private readonly AutoResetEvent evt;
            public event ComponentDataDelegate ComponentRegistered;
            public event ComponentModelDelegate ComponentModelCreated;
            public event EventHandler AddedAsChildKernel;
            public event EventHandler RemovedAsChildKernel;
            public event ComponentInstanceDelegate ComponentCreated;
            public event ComponentInstanceDelegate ComponentDestroyed;
            public event HandlerDelegate HandlerRegistered;
            public event HandlersChangedDelegate HandlersChanged;
            public event DependencyDelegate DependencyResolving;
            public event EventHandler RegistrationCompleted;
            public event ServiceDelegate EmptyCollectionResolving;

            public MyKernel(AutoResetEvent evt)
            {
                this.evt = evt;
            }

            public void Dispose()
            {
            }

            object IKernel.this[string key]
            {
                get { throw new NotImplementedException(); }
            }

            object IKernel.this[Type service]
            {
                get { throw new NotImplementedException(); }
            }

            public void AddComponent(string key, Type classType)
            {
                throw new NotImplementedException();
            }

            public void AddComponent(string key, Type classType, LifestyleType lifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent(string key, Type classType, LifestyleType lifestyle, bool overwriteLifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent(string key, Type serviceType, Type classType)
            {
                throw new NotImplementedException();
            }

            public void AddComponent(string key, Type serviceType, Type classType, LifestyleType lifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent(string key, Type serviceType, Type classType, LifestyleType lifestyle, bool overwriteLifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>()
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>(LifestyleType lifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>(LifestyleType lifestyle, bool overwriteLifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>(Type serviceType)
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>(Type serviceType, LifestyleType lifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponent<T>(Type serviceType, LifestyleType lifestyle, bool overwriteLifestyle)
            {
                throw new NotImplementedException();
            }

            public void AddComponentInstance<T>(object instance)
            {
                throw new NotImplementedException();
            }

            public void AddComponentInstance<T>(Type serviceType, object instance)
            {
                throw new NotImplementedException();
            }

            public void AddComponentInstance(string key, object instance)
            {
                throw new NotImplementedException();
            }

            public void AddComponentInstance(string key, Type serviceType, object instance)
            {
                throw new NotImplementedException();
            }

            public void AddComponentInstance(string key, Type serviceType, Type classType, object instance)
            {
                throw new NotImplementedException();
            }

            public void AddComponentWithExtendedProperties(string key, Type classType, IDictionary extendedProperties)
            {
                throw new NotImplementedException();
            }

            public void AddComponentWithExtendedProperties(string key, Type serviceType, Type classType, IDictionary extendedProperties)
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility(string key, IFacility facility)
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility<T>(string key) where T : IFacility, new()
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility<T>(string key, Action<T> onCreate) where T : IFacility, new()
            {
                throw new NotImplementedException();
            }

            public object Resolve(string key, object argumentsAsAnonymousType)
            {
                throw new NotImplementedException();
            }

            public object Resolve(string key, IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public IComponentModelBuilder ComponentModelBuilder { get; private set; }
            public IConfigurationStore ConfigurationStore { get; set; }
            public GraphNode[] GraphNodes { get; private set; }
            public IHandlerFactory HandlerFactory { get; private set; }
            public IKernel Parent { get; set; }
            public IProxyFactory ProxyFactory { get; set; }
            public IReleasePolicy ReleasePolicy { get; set; }
            public IDependencyResolver Resolver { get; private set; }

            public void AddChildKernel(IKernel kernel)
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility(IFacility facility)
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility<T>() where T : IFacility, new()
            {
                throw new NotImplementedException();
            }

            public IKernel AddFacility<T>(Action<T> onCreate) where T : IFacility, new()
            {
                throw new NotImplementedException();
            }

            public void AddHandlerSelector(IHandlerSelector selector)
            {
                throw new NotImplementedException();
            }

            public void AddHandlersFilter(IHandlersFilter filter)
            {
                throw new NotImplementedException();
            }

            public void AddSubSystem(string name, ISubSystem subsystem)
            {
                throw new NotImplementedException();
            }

            public IHandler[] GetAssignableHandlers(Type service)
            {
                throw new NotImplementedException();
            }

            public IFacility[] GetFacilities()
            {
                throw new NotImplementedException();
            }

            public IHandler GetHandler(string name)
            {
                throw new NotImplementedException();
            }

            public IHandler GetHandler(Type service)
            {
                throw new NotImplementedException();
            }

            public IHandler[] GetHandlers(Type service)
            {
                throw new NotImplementedException();
            }

            public ISubSystem GetSubSystem(string name)
            {
                throw new NotImplementedException();
            }

            public bool HasComponent(string name)
            {
                throw new NotImplementedException();
            }

            public int parallelCount = 0;
            private bool registered;

            public bool HasComponent(Type service)
            {
                return registered;
            }

            public IKernel Register(params IRegistration[] registrations)
            {
                Interlocked.Increment(ref parallelCount);
                evt.WaitOne();
                registered = true;
                Interlocked.CompareExchange(ref parallelCount, 0, 1);
                return null;
            }

            public void ReleaseComponent(object instance)
            {
                throw new NotImplementedException();
            }

            public void RemoveChildKernel(IKernel kernel)
            {
                throw new NotImplementedException();
            }

            public object Resolve(Type service)
            {
                throw new NotImplementedException();
            }

            public object Resolve(Type service, IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public object Resolve(Type service, object argumentsAsAnonymousType)
            {
                throw new NotImplementedException();
            }

            public object Resolve(string key, Type service)
            {
                throw new NotImplementedException();
            }

            public T Resolve<T>(IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public T Resolve<T>(object argumentsAsAnonymousType)
            {
                throw new NotImplementedException();
            }

            public T Resolve<T>()
            {
                return (T)(object)new DummyFactory();
            }

            public T Resolve<T>(string key)
            {
                throw new NotImplementedException();
            }

            public T Resolve<T>(string key, IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public object Resolve(string key, Type service, IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public Array ResolveAll(Type service)
            {
                throw new NotImplementedException();
            }

            public Array ResolveAll(Type service, IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public Array ResolveAll(Type service, object argumentsAsAnonymousType)
            {
                throw new NotImplementedException();
            }

            public TService[] ResolveAll<TService>()
            {
                throw new NotImplementedException();
            }

            public TService[] ResolveAll<TService>(IDictionary arguments)
            {
                throw new NotImplementedException();
            }

            public TService[] ResolveAll<TService>(object argumentsAsAnonymousType)
            {
                throw new NotImplementedException();
            }
        }

        private class DummyFactory : IPoolFactory
        {
            public IPool Create(int initialsize, int maxSize, IComponentActivator activator)
            {
                return new MyPool();
            }
        }

        private sealed class MyPool : IPool
        {
            public void Dispose()
            {
            }

            public bool Release(object instance)
            {
                throw new NotImplementedException();
            }

            public object Request(CreationContext context, Func<CreationContext, Burden> creationCallback)
            {
                throw new NotImplementedException();
            }
        }
```

**Comment by kkozmic** [comment]
Would you like to create a pull request with this fix?

**Comment by jonorossi** [comment]
Fixed in #289. Thanks @krinmik.
